lib/rules/curly.js

```diff
@@ -137,39 +140,59 @@ module.exports = {
          * @param {string} name The name to report.
          * @param {string} suffix Additional string to add to the end of a report.
          * @returns {void}
          * @private
          */
-        function reportExpectedBraceError(node, name, suffix) {
+        function reportExpectedBraceError(node, bodyNode, name, suffix) {
             context.report({
                 node,
                 loc: (name !== "else" ? node : getElseKeyword(node)).loc.start,
                 message: "Expected { after '{{name}}'{{suffix}}.",
                 data: {
                     name,
                     suffix: (suffix ? ` ${suffix}` : "")
-                }
+                },
+                fix: fixer => fixer.replaceText(bodyNode, `{${sourceCode.getText(bodyNode)}}`)
             });
         }
 
         /**
          * Reports "Unnecessary { after ..." error
          * @param {ASTNode} node The node to report.
+         * @param {ASTNode} bodyNode The block statement that is incorrectly surrounded by parens
          * @param {string} name The name to report.
          * @param {string} suffix Additional string to add to the end of a report.
          * @returns {void}
          * @private
          */
-        function reportUnnecessaryBraceError(node, name, suffix) {
+        function reportUnnecessaryBraceError(node, bodyNode, name, suffix) {
             context.report({
                 node,
                 loc: (name !== "else" ? node : getElseKeyword(node)).loc.start,
                 message: "Unnecessary { after '{{name}}'{{suffix}}.",
                 data: {
                     name,
                     suffix: (suffix ? ` ${suffix}` : "")
+                },
+                fix(fixer) {
+
+                    // `do while` expressions sometimes need a space to be inserted after `do`.
+                    // e.g. `do{foo()} while (bar)` should be corrected to `do foo() while (bar)`
+                    const needsPrecedingSpace = node.type === "DoWhileStatement" &&
+                        sourceCode.getTokenBefore(bodyNode).end === bodyNode.start &&
+                        esUtils.code.isIdentifierPartES6(sourceCode.getText(bodyNode).charCodeAt(1));
+
+                    // Insert a semicolon after the statement if it does not already have a semicolon, and if it's not the last
+                    // statement of the file. e.g. `if (foo) {bar()} baz()` gets fixed to `if (foo) bar(); baz()`. This is not a
+                    // perfect check; it will sometimes add semicolons when it doesn't actually have to, e.g.
+                    // `function foo() { if (foo) {bar()} }` will get fixed to `function foo() { if (foo) bar(); }`. However, the `semi`
+                    // rule should clean up afterwards if the semicolon is unnecessary and the user prefers no semicolons.
+                    const needsTrailingSemicolon = !/;\s*}$/.test(sourceCode.getText(bodyNode)) && sourceCode.getTokenAfter(bodyNode);
```

Comment:
I think I fixed that case; it now ignores `BlockStatement` nodes whose parent is a `FunctionExpression` or `ArrowFunctionExpression`.